internal/ingress/metric/collector/stats.go

```diff
@@ -0,0 +1,193 @@
+	RequestDuration float64 `json:"requestDuration"` // Metric
+
+	UpstreamName         string  `json:"upstreamName"`         // Label
+	UpstreamIP           string  `json:"upstreamIP"`           // Label
+	UpstreamResponseTime float64 `json:"upstreamResponseTime"` // Metric
+	UpstreamStatus       string  `json:"upstreamStatus"`       // Label
+
+	Namespace string `json:"namespace"` // Label
+	Ingress   string `json:"ingress"`   // Label
+	Service   string `json:"service"`   // Label
+}
+
+// StatsCollector stores prometheus metrics and ingress meta-data
+type StatsCollector struct {
+	upstreamResponseTime *prometheus.HistogramVec
+	requestDuration      *prometheus.HistogramVec
+	requestLength        *prometheus.HistogramVec
+	bytesSent            *prometheus.HistogramVec
+	listener             *net.UDPConn
+	ns                   string
+	watchClass           string
+	port                 int
+}
+
+// NewInstance creates a new StatsCollector instance
+func NewInstance(ns string, class string, port int) (*StatsCollector, error) {
+	sc := StatsCollector{}
+
+	listener, err := newUDPListener(port)
+
+	if err != nil {
+		return nil, err
+	}
+
+	sc.listener = listener
+	sc.ns = ns
+	sc.watchClass = class
+	sc.port = port
+
+	tags := []string{"host", "status", "remote_address", "remote_user", "protocol", "method", "path", "upstream_name", "upstream_ip", "upstream_status", "namespace", "ingress", "service"}
+
+	sc.upstreamResponseTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "upstream_response_time_seconds",
+			Help:      "The time spent on receiving the response from the upstream server",
+			Namespace: ns,
+		},
+		tags,
+	)
+
+	sc.requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "request_duration_seconds",
+			Help:      "The request processing time in seconds",
+			Namespace: ns,
+		},
+		tags,
+	)
+
+	sc.requestLength = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "request_length_bytes",
+			Help:      "The request length (including request line, header, and request body)",
+			Namespace: ns,
+			Buckets:   prometheus.LinearBuckets(10, 10, 10), // 10 buckets, each 10 bytes wide.
+		},
+		tags,
+	)
+
+	sc.bytesSent = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:      "bytes_sent",
+			Help:      "The the number of bytes sent to a client",
+			Namespace: ns,
+			Buckets:   prometheus.ExponentialBuckets(10, 10, 7), // 7 buckets, exponential factor of 10.
+		},
+		tags,
+	)
+
+	prometheus.MustRegister(sc.upstreamResponseTime)
+	prometheus.MustRegister(sc.requestDuration)
+	prometheus.MustRegister(sc.requestLength)
+	prometheus.MustRegister(sc.bytesSent)
+	return &sc, nil
+}
+
+func (sc *StatsCollector) handleMessage(msg []byte) {
+	glog.V(5).Infof("msg: %v", string(msg))
+
+	// Unmarshall bytes
+	var stats udpData
+	err := json.Unmarshal(msg, &stats)
+	if err != nil {
+		panic(err)
```

Comment:
This will panic only the collector or the controller?

Reply:
The controller.

Reply:
@fmejia97 please change that line with something like
```
glog.Errorf("Unexpected error deserializing JSON paylod: %v", err)
return
```

Reply:
👍